### PR TITLE
add NewStatus uplink command

### PR DESCRIPTION
**Pull request by DeadByElpy**
tests in next pull request

### CODE REVIEW

File: src/CommandBinaryBuffer.ts [hunk 1]
```diff
@@ -7,6 +7,24 @@ const DATE_BIT_SIZE = 5;
...
+     */
+    low: number | undefined
+
+    /**
+     * battery voltage value at hight consumption, in mV;
+     *
+     * 4095 === undefined
+     */
+    high: number | undefined
+}
```
Comment (DarkPark) [inline src/CommandBinaryBuffer.ts:28]:
```suggestion


```

File: src/constants/hardwareTypes.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
+// GAS RM on cpu STM32L15x - no longer in production, over 10,000 were produced.
```
Comment (DarkPark) [inline src/constants/hardwareTypes.ts:1]: should we add this to public interface?

File: src/commands/uplink/NewStatus.ts [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+interface INewStatusParameters {
+    software: IProduct,
+    hardware: IProduct,
+    data: IStatusBase
+}
+
+
+/**
+ * Uplink command.
+ *
```
Comment (DarkPark) [inline src/commands/uplink/NewStatus.ts:67]: whoud be nice to have an example here

File: src/commands/uplink/NewStatus.ts [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+    static readonly id = COMMAND_ID;
+
+    static readonly isUplink = true;
+
+    static readonly title = COMMAND_TITLE;
+
+    static fromBytes ( data: Uint8Array ): NewStatus {
+        const buffer = new CommandBinaryBuffer(data);
+
```
Comment (DarkPark) [inline src/commands/uplink/NewStatus.ts:83]:
```suggestion
```

File: src/commands/uplink/NewStatus.ts [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+            default:
+                throw new Error(`${this.getId()}: hardware type ${hardware.type} not supported`);
+        }
+
+        return new NewStatus({software, hardware, data: statusData});
+    }
+
+    toBytes (): Uint8Array {
+        const {software, hardware, data} = this.parameters;
+        const buffer = new CommandBinaryBuffer(COMMAND_BODY_MAX_SIZE, false);
```
Comment (DarkPark) [inline src/commands/uplink/NewStatus.ts:130]: in fromBytes CommandBinaryBuffer is created with LE

is it ok?

File: src/commands/uplink/NewStatus.ts [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+     * Remaining battery capacity
+     *
+     * 254 - 100%
+     * 255 - unknown
+     *
+     * @example
+     */
+    remindedBatteryCapacity: number | undefined
+
+    lastEventNumber: number
```
Comment (DarkPark) [inline src/commands/uplink/NewStatus.ts:37]: is it about sequence number?

File: src/commands/uplink/NewStatus.ts (hunk 3 above)
Comment (DeadByElpy) [reply src/commands/uplink/NewStatus.ts:67]: in next PR

File: src/constants/hardwareTypes.ts (hunk 2 above)
Comment (DeadByElpy) [reply src/constants/hardwareTypes.ts:1]: not now

File: src/commands/uplink/NewStatus.ts (hunk 6 above)
Comment (DeadByElpy) [reply src/commands/uplink/NewStatus.ts:37]: sequence number of event

File: src/CommandBinaryBuffer.ts (hunk 1 above)
Comment (DeadByElpy) [reply src/CommandBinaryBuffer.ts:28]: fixed

File: src/commands/uplink/NewStatus.ts (hunk 4 above)
Comment (DeadByElpy) [reply src/commands/uplink/NewStatus.ts:83]: fixed

File: src/commands/uplink/NewStatus.ts [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+                if ( statusData.remindedBatteryCapacity === UNKNOWN_CAPACITY ) {
+                    statusData.remindedBatteryCapacity = undefined;
+                }
+
+                break;
+
+            case hardwareTypes.MTXLORA:
+            case hardwareTypes.ELIMP:
+            default:
+                throw new Error(`${this.getId()}: hardware type ${hardware.type} not supported`);
```
[the remark this answers is not in the document]
Comment (DeadByElpy) [reply src/commands/uplink/NewStatus.ts:122]: fixed, in NewEvent

File: src/commands/uplink/NewStatus.ts (hunk 5 above)
Comment (DeadByElpy) [reply src/commands/uplink/NewStatus.ts:130]: fixed in all commands and tested

File: src/commands/uplink/NewStatus.ts (hunk 6 above)
Comment (DeadByElpy) [reply src/commands/uplink/NewStatus.ts:37]: fixed

File: src/commands/downlink/NewStatus.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+    static readonly id = COMMAND_ID;
+
+    static readonly isUplink = false;
+
+    static readonly title = COMMAND_TITLE;
+
+    // data - only body (without header)
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    static fromBytes ( _data: Uint8Array ) {
```
Comment (DarkPark) [inline src/commands/downlink/NewStatus.ts:21]:
```suggestion
    static fromBytes () {
```
?

File: src/commands/downlink/NewStatus.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        return new NewStatus();
+    }
+
+    // eslint-disable-next-line class-methods-use-this
+    toBytes (): Uint8Array {
+        return Command.toBytes(COMMAND_ID, new Uint8Array());
```
Comment (DarkPark) [inline src/commands/downlink/NewStatus.ts:27]:
```suggestion
        return Command.toBytes(COMMAND_ID);
```

File: src/commands/uplink/NewStatus.ts [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+const UNKNOWN_RESISTANT = 65535;
+
+// max battery capacity, 254 - 100%
+const UNKNOWN_BATTERY_CAPACITY = 255;
+
+
+/**
+ * NewStatus command parameters
+ */
+interface INewStatusParameters {
```
Comment (DarkPark) [inline src/commands/uplink/NewStatus.ts:51]: let's put all interfaces together

File: src/commands/downlink/NewStatus.ts [hunk 11]
```diff
@@ -0,0 +1,32 @@
+import Command from '../../Command.js';
+
+
+const COMMAND_ID = 0x14;
+const COMMAND_TITLE = 'NEW_STATUS';
+
+
+class NewStatus extends Command {
```
Comment (DarkPark) [inline src/commands/downlink/NewStatus.ts:8]: description and example?

File: src/commands/downlink/NewStatus.ts (hunk 8 above)
Comment (DeadByElpy) [reply src/commands/downlink/NewStatus.ts]: fixed

File: src/commands/downlink/NewStatus.ts (hunk 9 above)
Comment (DeadByElpy) [reply src/commands/downlink/NewStatus.ts]: fixed

File: src/commands/uplink/NewStatus.ts (hunk 10 above)
Comment (DeadByElpy) [reply src/commands/uplink/NewStatus.ts:41]: reworked in all files, utils/message.ts and utils/getHexFromBytes.ts were already structured in this way

File: src/commands/uplink/NewStatus.ts [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+     * undefined - unknown capacity
+     *
+     */
+    remindedBatteryCapacity: number | undefined
+
+    lastEventSequenceNumber: number
+}
+
+
+const COMMAND_ID = 0x14;
```
[the remark this answers is not in the document]
Comment (DeadByElpy) [reply src/commands/uplink/NewStatus.ts]: reworked in all files, utils/message.ts and utils/getHexFromBytes.ts were already structured in this way

File: src/commands/downlink/NewStatus.ts (hunk 11 above)
Comment (DeadByElpy) [reply src/commands/downlink/NewStatus.ts:23]: added

File: tests/commands.test.ts [hunk 13]
```diff
@@ -54,6 +54,16 @@ const downlinkCommands: TCommandList = [
             body: '',
             lrc: '4c'
         }
+    },
+    {
+        constructor: downlink.NewStatus,
+        name: 'downlink command 0x14:NEW_STATUS',
```
Comment (DarkPark) [inline tests/commands.test.ts:60]: let's keep it sorted

Comment (DeadByElpy) [reply tests/commands.test.ts]: done

**Review by DarkPark — APPROVED** [review]

**Pull request merged**